server/peers.py

```diff
@@ -560,9 +560,11 @@ def retry_peer(self, peer, port_pairs):
             create_connection = self.proxy.create_connection
         else:
             create_connection = self.loop.create_connection
+            
+        laddr=(self.env.host, None) if self.env.host else None    
```

Comment:
> Do you not need to set the port in the tuple? I can't test as I don't have this setup.  

AFAIK source port selection is normally done by the operating system, as each outgoing connection does need its own source port. So with many connections it would be very problematic to choose a port beforehand which is not occupied.  
Python docs state that you can use None to pass null to system layer: https://docs.python.org/3/library/socket.html#socket.getaddrinfo  

That said, it works for me in both cases (with and without HOST= set). But it should certainly receive some testing, especially in other (non-Linux) environments.

Regarding the other changes, I'm no programmer and my Python knowledge is minimal. But I'll do my best ;)